pkg/services/live/push/push.go

```diff
@@ -58,25 +80,97 @@ func (g *Gateway) IsEnabled() bool {
 	return g.Cfg.IsLiveEnabled() // turn on when Live on for now.
 }
 
-func (g *Gateway) Handle(ctx *models.ReqContext) {
-	streamID := ctx.Params(":streamId")
+func extractKey(req *http.Request) string {
+	header := req.Header.Get("Authorization")
+	parts := strings.SplitN(header, " ", 2)
+	var keyString string
+	if len(parts) == 2 && parts[0] == "Bearer" {
+		keyString = parts[1]
+	}
+	return keyString
+}
+
+func (g *Gateway) isAuthenticated(req *http.Request) (bool, error) {
+	keyString := extractKey(req)
+	if keyString == "" {
+		return false, nil
+	}
+	g.keyMu.RLock()
+	existingKey, ok := g.keyCache[keyString]
+	g.keyMu.RUnlock()
+	if !ok {
+		// TODO: share the code below with middleware package.
+		decoded, err := apikeygen.Decode(keyString)
+		if err != nil {
+			return false, nil
+		}
+		keyQuery := models.GetApiKeyByNameQuery{KeyName: decoded.Name, OrgId: decoded.OrgId}
+		if err := bus.Dispatch(&keyQuery); err != nil {
+			if errors.Is(err, models.ErrInvalidApiKey) {
+				return false, nil
+			}
+			return false, err
+		}
+		apikey := keyQuery.Result
+		// validate api key.
+		isValid, err := apikeygen.IsValid(decoded, apikey.Key)
+		if err != nil {
+			return false, nil
+		}
+		if !isValid {
+			return false, nil
+		}
 
+		now := time.Now()
+
+		if apikey.Expires != nil && *apikey.Expires <= now.Unix() {
+			return false, nil
+		}
+		g.keyMu.Lock()
+		g.keyCache[keyString] = CachedKey{CheckedAt: now, IsValid: true}
+		g.keyMu.Unlock()
+	} else {
+		if !existingKey.IsValid {
+			g.keyMu.Lock()
+			delete(g.keyCache, keyString)
+			g.keyMu.Unlock()
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (g *Gateway) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	ok, err := g.isAuthenticated(req)
+	if err != nil {
+		logger.Error("Error authenticating push request", "error", err)
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if !ok {
+		rw.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+	g.handleStreamPush(rw, req)
+}
+
+func (g *Gateway) handleStreamPush(rw http.ResponseWriter, req *http.Request) {
+	// TODO: properly extract streamID.
+	streamID := "telegraf"
 	stream, err := g.GrafanaLive.ManagedStreamRunner.GetOrCreateStream(streamID)
 	if err != nil {
 		logger.Error("Error getting stream", "error", err)
-		ctx.Resp.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
 	// TODO Grafana 8: decide which formats to use or keep all.
-	urlValues := ctx.Req.URL.Query()
+	urlValues := req.URL.Query()
 	frameFormat := pushurl.FrameFormatFromValues(urlValues)
 	stableSchema := pushurl.StableSchemaFromValues(urlValues)
 
-	body, err := ctx.Req.Body().Bytes()
+	body, err := ioutil.ReadAll(req.Body)
```

Comment:
`ioutil` is deprecated:
```suggestion
	body, err := io.ReadAll(req.Body)
```